api/v1beta1/clusterclass_types.go

```diff
@@ -542,7 +545,127 @@ type JSONSchemaProps struct {
 	// NOTE: Can be set for all types.
 	// +optional
 	Default *apiextensionsv1.JSON `json:"default,omitempty"`
-}
+
+	// XValidations describes a list of validation rules written in the CEL expression language.
+	// +optional
+	XValidations ValidationRules `json:"x-kubernetes-validations,omitempty"`
+}
+
+// ValidationRules describes a list of validation rules written in the CEL expression language.
+// +listType=map
+// +listMapKey=rule
+type ValidationRules []ValidationRule
+
+// ValidationRule describes a validation rule written in the CEL expression language.
+type ValidationRule struct {
```

Comment:
Q: I wonder if the feature is entirely implemted as documented, e.g. details like this:

```go
	// If messageExpression results in a runtime error, the runtime error is logged, and the validation failure message is produced
	// as if the messageExpression field were unset. If messageExpression evaluates to an empty string, a string with only spaces, or a string
	// that contains line breaks, then the validation failure message will also be produced as if the messageExpression field were unset, and
	// the fact that messageExpression produced an empty string/string with only spaces/string with line breaks will be logged.
```

Also wondering if all fields are completely validated (or how far the validation goes, not sure what the k/k library does for us)

Reply:
I have added field validation now for x-validation-rules.

The k/k library does do the validation and expression output for us, there is a unit test in this PR for that.